syft/formats.go

```diff
@@ -0,0 +1,77 @@
+package syft
+
+import (
+	"bytes"
+
+	"github.com/anchore/syft/internal/formats/cyclonedx13json"
+	"github.com/anchore/syft/internal/formats/cyclonedx13xml"
+	"github.com/anchore/syft/internal/formats/spdx22json"
+	"github.com/anchore/syft/internal/formats/spdx22tagvalue"
+	"github.com/anchore/syft/internal/formats/syftjson"
+	"github.com/anchore/syft/internal/formats/table"
+	"github.com/anchore/syft/internal/formats/text"
+	"github.com/anchore/syft/syft/sbom"
+)
+
+const (
+	JSONFormatOption          FormatOption = "json"
+	TextFormatOption          FormatOption = "text"
+	TableFormatOption         FormatOption = "table"
+	CycloneDxXMLFormatOption  FormatOption = "cyclonedx-xml"
+	CycloneDxJSONFormatOption FormatOption = "cyclonedx-json"
+	SPDXTagValueFormatOption  FormatOption = "spdx-tag-value"
+	SPDXJSONFormatOption      FormatOption = "spdx-json"
+)
+
+var formats []sbom.Format
+
+func init() {
+	formats = []sbom.Format{
+		syftjson.Format(string(JSONFormatOption)),
+		cyclonedx13xml.Format(string(CycloneDxXMLFormatOption)),
+		cyclonedx13json.Format(string(CycloneDxJSONFormatOption)),
+		spdx22tagvalue.Format(string(SPDXTagValueFormatOption)),
+		spdx22json.Format(string(SPDXJSONFormatOption)),
+		table.Format(string(TableFormatOption)),
+		text.Format(string(TextFormatOption)),
+	}
+}
+
+type FormatOption string
+
+func FormatOptions() []FormatOption {
+	return []FormatOption{
+		JSONFormatOption,
+		TextFormatOption,
+		TableFormatOption,
+		CycloneDxXMLFormatOption,
+		CycloneDxJSONFormatOption,
+		SPDXTagValueFormatOption,
+		SPDXJSONFormatOption,
+	}
+}
+
+func FormatByOption(option FormatOption) sbom.Format {
+	return FormatByName(string(option))
+}
+
+func FormatByName(name string) sbom.Format {
+	for _, f := range formats {
+		for _, formatName := range f.Names() {
+			if formatName == name {
+				return f
+			}
+		}
+	}
+	return nil
+}
+
+func IdentifyFormat(by []byte) sbom.Format {
+	for _, f := range formats {
+		if err := f.Validate(bytes.NewReader(by)); err != nil {
```

Comment:
Nothing that will hold up this PR or even is relevant for the change here; just noting I think we we should revisit using Validate to identify a format -- when an error occurs, in what might be an otherwise valid file the user never will get notified what is wrong, this has come up just recently when someone tried to use SPDX as input but the file was invalid...

Reply:
A validator is meant to drive this identify function, and not necessarily validate the entire document. Take for instance the syft json validator:
```golang
func validator(reader io.Reader) error {
	type Document struct {
		Schema model.Schema `json:"schema"`
	}

	dec := json.NewDecoder(reader)

	var doc Document
	err := dec.Decode(&doc)
	if err != nil {
		return fmt.Errorf("unable to decode: %w", err)
	}

	// note: we accept all schema versions
	if strings.Contains(doc.Schema.URL, "anchore/syft") {
		return nil
	}
	return fmt.Errorf("could not extract syft schema")
}

```
If this is invalid json, we bail. If there is no schema block, we bail. If the schema URL doesn't reference anchore/syft, we bail. No other fields are referenced... and this is intentional.

Per https://github.com/anchore/syft/blob/main/syft/format/validator.go#L4-L12 a `Validator` is intentionally scoped as small as possible (for now).

> I think we we should revisit using Validate to identify a format 

I agree. I think maybe `Validate` should be changed to a name that more clearly indicates it's being used to identify the format. Maybe something like `IdentifiesAs([]bytes) bool` would be better (in which case this is an easy change to the format interface). We can later add full-format validators and use `Validate([]byte) error` for that.

Agreed on maybe a follow up PR.